### ISSUE TITLE
Regarding Slow Prediction Speed with TabPFN on Large-Scale Datasets (10 Million Samples)

### ISSUE THREAD

**Issue by zhitingjiang** [PriorLabs/tabpfn-extensions#49]
Dear TabPFN Team,

First, I want to express my sincere appreciation for your groundbreaking work. The model's performance is truly remarkable, and I'm currently applying it to ocean remote sensing image inversion tasks.

My dataset contains over 10,000 samples with 7 feature parameters (53,802 training samples and 23,061 validation samples). Following your recommendations, I adopted the RF-Preprocessing strategy which demonstrated excellent accuracy despite the substantial training time required - this outcome was particularly gratifying.

However, I'm encountering significant efficiency challenges during product generation. For our operational 4500×4100 grid products (approximately 10 million valid inputs after removing null values), the prediction phase becomes prohibitively slow. Even with batch processing of 10,000 samples per iteration (requiring >1000 cycles), each iteration takes 1-2 minutes on my system configured with:

- 128GB RAM
- NVIDIA RTX 4080 GPU

I would greatly appreciate your insights regarding:

- Whether this performance is hardware-related or potentially suboptimal in implementation

- Recommended optimization strategies for large-scale inference

Attached are code snippets for your reference:

**[Model Training Code]**
```
from sklearn.metrics import mean_squared_error, r2_score
import pandas as pd
from tabpfn_extensions.rf_pfn import RandomForestTabPFNRegressor
from tabpfn import TabPFNRegressor
import joblib
import numpy as np

Version = "20250320"
Disk = "G:\\"
Algorithm = "TabPFN"

# 数据加载
data_T = pd.read_csv(f"{Disk}Block1\\Code\\New_Model_Test\\TabPFN\\{Version}\\ECS_pCO2_Training_spacetime.csv", 
                    index_col=0, encoding='gb2312')
data_V = pd.read_csv(f"{Disk}Block1\\Code\\New_Model_Test\\TabPFN\\{Version}\\ECS_pCO2_Validation_spacetime.csv",
                    index_col=0, encoding='gb2312')

# 特征工程
X_train = data_T.iloc[:, 4:-1]
y_train = data_T.iloc[:, -1]
index_T = data_T.index
X_test = data_V.iloc[:, 4:-1]
y_test = data_V.iloc[:, -1]
index_V = data_V.index

# 初始化并训练模型
reg_base = TabPFNRegressor(
    ignore_pretraining_limits=True,
    inference_config = {"SUBSAMPLE_SAMPLES": 10000}
)

model = RandomForestTabPFNRegressor(
    tabpfn=reg_base,
    verbose=0,
    max_predict_time=60,
    fit_nodes=True,
    adaptive_tree=True,
)

print("开始训练模型...")
model.fit(X_train, y_train)

# Predict on the train set
predictions = model.predict(X_train)

# Evaluate the model
rmse = np.sqrt(mean_squared_error(y_train, predictions))
r2 = r2_score(y_train, predictions)

print("Root Mean Squared Error (RMSE):", rmse)
print("R² Score:", r2)

#将预测结果保存到文件中
result_Training = pd.DataFrame(index_T)
result_Training['pCO2'] = predictions
result_Training.to_csv(f"{Disk}Block1\\Code\\New_Model_Test\\TabPFN\\{Version}\\ECS_Predict_pCO2_Training_{Algorithm}.csv", encoding='gb2312', index=False)

# Predict on the test set
predictions = model.predict(X_test)

# Evaluate the model
rmse = np.sqrt(mean_squared_error(y_test, predictions))
r2 = r2_score(y_test, predictions)

print("Root Mean Squared Error (RMSE):", rmse)
print("R² Score:", r2)

#将预测结果保存到文件中
result_Validation = pd.DataFrame(index_V)
result_Validation['pCO2'] = predictions
result_Validation.to_csv(f"{Disk}Block1\\Code\\New_Model_Test\\TabPFN\\{Version}\\ECS_Predict_pCO2_Validation_{Algorithm}.csv", encoding='gb2312', index=False)

# 模型保存
save_path = f"{Disk}Block1\\Code\\New_Model_Test\\TabPFN\\{Version}\\rf_tabpfn_model.pkl"
joblib.dump(model, save_path)
print(f"\n模型已保存至: {save_path}")
```

**[Production Code]**
```
import numpy as np
import netCDF4 as nc
import joblib
from netCDF4 import Dataset
from datetime import datetime
import calendar
import pandas as pd
import os

Version = "20250320"  # 使用TabPFN模型对应的版本号

# range_lon = slice(2100, 2401)
# range_lat = slice(1299, 1700)

# 加载训练好的TabPFN模型
model_path = f"G:\\Block1\\Code\\New_Model_Test\\TabPFN\\{Version}\\rf_tabpfn_model.pkl"
tabpfn_model = joblib.load(model_path)

years = [str(year) for year in range(2003, 2004)]
months = [str(month).zfill(2) for month in range(7,8)]

for year in years:
    for month in months:
        print(f"Processing {year}-{month}...")
        
        # 数据读取与预处理（保持原有逻辑）
        DDD1 = str((datetime(int(year),int(month),1)-datetime(int(year),1,1)).days+1).zfill(3)
        DDD2 = str(datetime(int(year), int(month), calendar.monthrange(int(year), int(month))[1]).timetuple().tm_yday).zfill(3)
        
        # 读取各环境变量数据
        filename_SST = f"E:\\Carbon_Project\\CS_Input\\nearest\\SST_1km\\CS-oisst-avhrr-v02r01-monthly-1km.{year}{month}.nc"
        data_SST = nc.Dataset(filename_SST)
        SST = data_SST.variables['SST'][:].data
        lon = data_SST.variables['lon'][:]
        lat = data_SST.variables['lat'][:]
        SST[SST == -999] = np.nan

        filename_SSTlat = f"E:\\Carbon_Project\\CS_Input\\nearest\\UI_1km\\Chinasea_UI_OISSTv2.1_{year}{month}_MO_1km_099140.nc"
        data_SSTlat = nc.Dataset(filename_SSTlat)
        SST_lat = np.transpose(data_SSTlat.variables['UI'][:].data)
        # SST_lat[SST_lat==-32767] = np.nan
        
        filename_Chla = f"E:\\Carbon_Project\\CS_Input\\nearest\\Chla_1km\\CS_ESACCI-OC-L3S-CHLOR_A-MERGED-1M_MONTHLY_1km_GEO_PML_OCx-{year}{month}-fv6.0.nc"
        data_Chla = nc.Dataset(filename_Chla)
        Chla = data_Chla.variables['Chla'][:].data
        Chla[Chla==-999] = np.nan
        
        filename_Rrs443 = f"E:\\Carbon_Project\\CS_Input\\nearest\\Rrs_443_1km\\CS_ESACCI-OC-L3S-RRS443-MERGED-1M_MONTHLY_4km_GEO_PML_RRS-{year}{month}-fv6.0.nc"
        data_Rrs443 = nc.Dataset(filename_Rrs443)
        Rrs_443 = data_Rrs443.variables['Rrs_443'][:].data
        Rrs_443[Rrs_443==-999] = np.nan
        
        filename_Rrs490 = f"E:\\Carbon_Project\\CS_Input\\nearest\\Rrs_490_1km\\CS_ESACCI-OC-L3S-RRS490-MERGED-1M_MONTHLY_4km_GEO_PML_RRS-{year}{month}-fv6.0.nc"
        data_Rrs490 = nc.Dataset(filename_Rrs490)
        Rrs_490 = data_Rrs490.variables['Rrs_490'][:].data
        Rrs_490[Rrs_490==-999] = np.nan
        
        filename_Rrs560 = f"E:\\Carbon_Project\\CS_Input\\nearest\\Rrs_560_1km\\CS_ESACCI-OC-L3S-RRS560-MERGED-1M_MONTHLY_4km_GEO_PML_RRS-{year}{month}-fv6.0.nc"
        data_Rrs560 = nc.Dataset(filename_Rrs560)
        Rrs_560 = data_Rrs560.variables['Rrs_560'][:].data
        Rrs_560[Rrs_560==-999] = np.nan
        
        filename_pCO2thermal = f"E:\\Carbon_Project\\CS_Input\\nearest\\pCO2thermal_1km\\CS_pco2therm_monthly_1km_{year}{month}.nc"
        data_pCO2thermal = nc.Dataset(filename_pCO2thermal)
        pCO2_thermal = data_pCO2thermal.variables['pCO2thermal'][:].data
        pCO2_thermal[pCO2_thermal==-999] = np.nan

        # 数据整形
        input_shape = (4500*4100, 1)
        SST = SST.reshape(*input_shape)
        SST_lat = SST_lat.reshape(*input_shape)
        Chla = Chla.reshape(*input_shape)
        Rrs_443 = Rrs_443.reshape(*input_shape)
        Rrs_490 = Rrs_490.reshape(*input_shape)
        Rrs_560 = Rrs_560.reshape(*input_shape)
        pCO2_thermal = pCO2_thermal.reshape(*input_shape)

        # 构建特征矩阵
        features = np.concatenate(
            (SST, SST_lat, Chla, Rrs_443, Rrs_490, Rrs_560, pCO2_thermal),
            axis=1
        )

        # 处理缺失值
        nan_mask = np.isnan(features).any(axis=1)
        valid_indices = ~nan_mask
        
        # 分块预测（避免内存不足）
        chunk_size = 10000
        # 初始化预测结果数组
        Y_pred = np.full(features.shape[0], np.nan)
        
        # 获取有效索引的物理位置
        valid_indices_flat = np.where(valid_indices)[0]  # 形如 [3,5,8,...] 的具体索引
        
        # 分块处理有效索引
        for i in range(0, len(valid_indices_flat), chunk_size):
            print(f"Processing chunk {i//chunk_size + 1}/{(len(valid_indices_flat)-1)//chunk_size + 1}")
            
            # 获取当前块的全局索引
            chunk_indices = valid_indices_flat[i:i+chunk_size]
            
            # 提取对应数据块
            chunk_data = features[chunk_indices]
            
            # 预测
            Y_pred_chunk = tabpfn_model.predict(chunk_data)
            
            # 直接通过物理索引赋值
            Y_pred[chunk_indices] = Y_pred_chunk

        # 结果整形
        Y_pred = Y_pred.reshape(4500, 4100)

        # 输出文件设置
        filename_output = f"G:\\Block1\\Code\\New_Model_Test\\TabPFN\\{Version}\\Product\\SIO_{year}{month}_pCO2_1km_ECS_TabPFN_{Version}.nc"
        
        # 创建NetCDF文件（保持原有格式）
        with Dataset(filename_output, 'w', format='NETCDF4') as pCO2_DataSet:
            # 维度定义
            pCO2_DataSet.createDimension('lon', 4500)
            pCO2_DataSet.createDimension('lat', 4100)
            
            # 变量定义
            pCO2_var = pCO2_DataSet.createVariable(
                'pCO2', 'f4', ('lat','lon'), 
                zlib=True, complevel=9, fill_value=-9999
            )
            pCO2_var.long_name = 'Partial Pressure of Carbon Dioxide'
            pCO2_var.units = 'μatm'
            pCO2_var.coordinate = 'lon lat'
            pCO2_var.grid_mapping = 'lambert_conformal_conic'
            
            # 坐标变量
            lon_var = pCO2_DataSet.createVariable('lon', 'f4', ('lon',))
            lon_var[:] = lon
            lat_var = pCO2_DataSet.createVariable('lat', 'f4', ('lat',))
            lat_var[:] = lat
            
            # 全局属性
            pCO2_DataSet.creation_date = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            pCO2_DataSet.processing_level = 'L3 Mapped'
            pCO2_DataSet.cdm_data_type = 'grid'
            pCO2_DataSet.temporal_range = 'monthly'
            pCO2_DataSet.model_info = f"RandomForestTabPFNRegressor v{Version}"
            
            # 写入数据
            pCO2_var[:, :] = Y_pred

print("所有月份处理完成")
```

Your expert guidance on this matter would be invaluable for both my research and practical applications.

Best regards,
Zhiting Jiang
Shanghai Jiao Tong University

**Comment by zhitingjiang** [comment]
Despite upgrading preprocessing.py to the latest version (supporting >10k samples), processing a single 4500×4100 grid still requires 20 hours, and completing 316-month data analysis would take over 8 months. 

I think this model is likely to be used in a wide variety of industries, however, data volume and computational efficiency is a topic that cannot be avoided. I hope the team can give me some advice. Thank you. Thank you.

**Comment by newlife0007** [comment]
I ran into the same issue. Did you find a solution? I saved the fitted model using joblib.dump, but the file turned out to be extremely large—tens of gigabytes—and the prediction is very slow.

**Comment by noahho** [comment]
Thanks for the input! TabPFN is currently made for datasets with size smaller than 10000 x 500. We are working on building a version that will support much larger data, stay tuned!
